### PR TITLE
Add new CSRF manager for unit testing purposes

### LINKED ISSUE #12159: Perform CSRF check not within request class
Currently the CSRF check is performed within the request class (`OC\AppFramework\Http:: passesCSRFCheck`) which is very weird and should be moved to a better place such as https://github.com/owncloud/core/blob/559b34e53d1adda66e12b3bd0c7a7459c9304a5c/lib/private/appframework/middleware/security/securitymiddleware.php#L114

### PR DESCRIPTION
**Pull request by LukasReschke**
This adds a new CSRF manager for unit testing purposes, it's interface is based upon https://github.com/symfony/security-csrf. Due to some of our required custom changes it is however not possible to use the Symfony component directly. Most of this are unit test adjustments.

Furthermore, this regenerates the CSRF token upon login because otherwise somebody else might be able to note down the CSRF token before login on a shared computer.

cc @rullzer @MorrisJobke @owncloud/security 

<hr/> 

Fixes https://github.com/owncloud/core/issues/12159

### CODE REVIEW

File: lib/private/security/csrf/tokenstorage/storageinterface.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+namespace OC\Security\CSRF\TokenStorage;
+
+/**
+ * Interface StorageInterface provides the interface required for token storage.
+ *
+ * @package OC\Security\CSRF\TokenStorage
+ */
+interface StorageInterface {
```
Comment (DeepDiver1975) [inline lib/private/security/csrf/tokenstorage/storageinterface.php]: do we expect to have multiple implementation here?

File: lib/public/util.php [hunk 2]
```diff
@@ -479,19 +479,22 @@ static public function emitHook( $signalclass, $signalname, $params = array()) {
 		return(\OC_Hook::emit( $signalclass, $signalname, $params ));
 	}
 
+	private static $token = '';
```
Comment (DeepDiver1975) [inline lib/public/util.php]: :see_no_evil: 

please add phpdoc

**Comment by DeepDiver1975** [comment]
a lot of stuff foe just one little tiny token ....

**Comment by LukasReschke** [comment]
> a lot of stuff foe just one little tiny token ....

To be fair, most of it is simply boiler plate code (e.g. copyright header, class declaration, PHPDocs) and in the end it's simply moving static code out of some place and putting it in a unit-tested other place. As well as giving additional functionality such as invalidating a token (which was actually the reason for the change).

The alternative would have been adding more static untested code. In fact, I'm tempted to say the actual logic has not really increased at all.

Other comments have been addressed.

**Comment by rullzer** [comment]
Looks a lot cleaner and yay for more non-static code!

I did some tests and everything still works and the code did not burn down my house.

:+1:

**Comment by MorrisJobke** [comment]
Code looks good and it works :+1:

**Pull request merged**